pkg/controller/application/app_controller.go

```diff
@@ -0,0 +1,295 @@
+	return func(obj interface{}) {
+		if event, ok := obj.(events.ApplicationEvent); ok {
+			if events.AppStateChange == event.GetEvent() {
+				if shimEvent, ok := event.(shimcache.ApplicationStatusChangeEvent); ok {
+					appID := event.GetApplicationID()
+					log.Logger.Info("Status Change callback received",
+						zap.String("app id", appID),
+						zap.String("new status", shimEvent.State))
+					var app = appMgr.amProtocol.GetApplication(appID).(*shimcache.Application)
+					appCRD, err := appMgr.controller.lister.Applications(app.GetTags()[constants.AppTagNamespace]).Get(appID)
+					if err != nil {
+						log.Logger.Warn("Failed to query app CRD for status update",
+							zap.String("Application ID", appID),
+							zap.Error(err))
+						return
+					}
+					crdState := convertShimAppStateToAppCRDState(shimEvent.State)
+					if crdState != "Undefined" {
+						appMgr.updateAppCRDStatus(appCRD, crdState)
+					} else {
+						log.Logger.Error("Invalid status, skip saving it",
+							zap.String("App id", appID))
+					}
+				}
+			}
+		}
+	}
+}
+
+type Controller struct {
+	kubeClient       kubernetes.Interface
+	extensionsClient apiextensionsclient.Interface
+	appClient        applicationclient.Interface
+	informer         cache.SharedIndexInformer
+	lister           applister.ApplicationLister
+}
+
+func NewController(appMgr *AppManager) *Controller {
+	kubeconfig := os.Getenv("KUBECONFIG")
+	config, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
+	if err != nil {
+		log.Logger.Fatal(err.Error())
+	}
+
+	kubeClient := kubernetes.NewForConfigOrDie(config)
+	apiextensionsClient := apiextensionsclient.NewForConfigOrDie(config)
+	appClient := applicationclient.NewForConfigOrDie(config)
+
+	informerFactory := appinformers.NewSharedInformerFactory(appClient, time.Minute*1)
+	informer := informerFactory.Apache().V1alpha1().Applications()
+	informer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    appMgr.addApp,
+		UpdateFunc: appMgr.updateApp,
+		DeleteFunc: appMgr.deleteApp,
+	})
+	informerFactory.Start(wait.NeverStop)
+
+	utilruntime.Must(appv1.AddToScheme(appscheme.Scheme))
+	return &Controller{
+		kubeClient:       kubeClient,
+		extensionsClient: apiextensionsClient,
+		appClient:        appClient,
+		informer:         informer.Informer(),
+		lister:           informer.Lister(),
+	}
+}
+
+func (c *Controller) Run() {
+	//TODO: handle the case when the crd is not registered: register it programatically
+	defer utilruntime.HandleCrash()
+
+	log.Logger.Info("Waiting cache to be synced.")
+	// Handle timeout for syncing.
+	timeout := time.NewTimer(time.Second * 30)
+	timeoutCh := make(chan struct{})
+	go func() {
+		<-timeout.C
+		timeoutCh <- struct{}{}
+	}()
+	if ok := cache.WaitForCacheSync(timeoutCh, c.informer.HasSynced); !ok {
+		log.Logger.Error("Timeout expired during waiting for caches to sync. Skip starting the application controller")
```

Comment:
this log is wrong, it is not skipping the controller start, everything has been started. this only means the controller is running without the up-to-date sync'd cache.